Implementing Ahocorasick Search.

35C4n0r opened the pull request:
This PR Implements the pure python variant of pyahocorasick, and heavily reuses code from ahocorapy with minor additions.
Also added methods for pickling and unpickling.

- [x] add_word  `used in SCTK`
- [x] get `used in SCTK`
- [x] clear
- [ ] dump
- [x] exists
- [x] find_all
- [ ] get_stats
- [x] items
- [x] make_automaton `used in SCTK`
- [x] iter `used in SCTK`
- [ ] iter_long
- [x] keys
- [ ] longest_prefix
- [ ] match
- [ ] pop
- [ ] remove_word
- [ ] save
- [x] values `used in SCTK`

35C4n0r committed:
Add: initial commit for the ahocode
Added basic code for Automaton creation and trie search.

Signed-off-by: 35C4n0r <jaykumar20march@gmail.com>

35C4n0r committed:
Add: added the remaining methods that are used in SCTK.
- Added method iter().
- Added methods keys() and values().
- added __getstate__() and __setstate__() for pickling.

Signed-off-by: 35C4n0r <jaykumar20march@gmail.com>

JonoYang commented on src/ahocode.py:
```diff
@@ -0,0 +1,341 @@
+AHOCORASICK = 2
```
Add this at the top of the file:
```
#
# Copyright (c) nexB Inc. and others. All rights reserved.
# ahocode is a trademark of nexB Inc.
# SPDX-License-Identifier: Apache-2.0
# See http://www.apache.org/licenses/LICENSE-2.0 for the license text.
# See https://github.com/nexB/ahocode for support or download.
# See https://aboutcode.org for more information about nexB OSS projects.
#
```

JonoYang commented on src/ahocode.py:
```diff
@@ -0,0 +1,341 @@
+
+    def __str__(self):
+        transitions_as_string = ','.join(
+            ['{0} -> {1}'.format(key, value.identifier) for key, value in self.transitions.items()])
```
I don't think you need to have your list comprehension in a list if you're passing it directly into `.join()`.
```suggestion
            '{0} -> {1}'.format(key, value.identifier) for key, value in self.transitions.items())

```

JonoYang commented on src/ahocode.py:
```diff
@@ -0,0 +1,341 @@
+        zero_state = self._zero_state
+        current_state = zero_state
+        offset = 0
+        if (start and end) is not None:
```
I notice that you use the `is` keyword when doing value comparison. You should double check where you use this because `is` and `==` do different things when comparing variables: https://stackoverflow.com/a/133024

JonoYang commented on src/ahocode.py:
```diff
@@ -0,0 +1,341 @@
+        return current_state.value
+
+    def values(self, prefix=None, wildcard=None, how=None):
+        for key, value in list(self.items()):
```
I feel like you could get away with not wrapping `self.items()` in a list.
```suggestion
        for key, value in self.items():

```

JonoYang commented on src/ahocode.py:
```diff
@@ -0,0 +1,341 @@
+
+    def __getstate__(self):
+        state_list = [None] * self._counter
+        todo_list = [self._zero_state]
```
I think a more descriptive variable can be used, instead of `todo_list`

JonoYang reviewed (changes requested):
@35C4n0r 

Great start! I've left some comments for you. In particular, be careful when you use `is` when comparing variables and `==` should be used when you are comparing variable values.

Also, update the `setup.cfg` and `README.rst` for this new project.

35C4n0r committed:
ref: refactored variables

Signed-off-by: 35C4n0r <jaykumar20march@gmail.com>

35C4n0r committed:
fix: fixed the keys and values method

- Fixed the logic for keys() and values() method, the logics were earlier interchanged.

Signed-off-by: 35C4n0r <jaykumar20march@gmail.com>

35C4n0r committed:
add: added new test files

- Added test.py for basic methods, such as add_word(), make_automaton(), iter(), find_all(), keys() and values().
- Updated setup.cfg

Signed-off-by: 35C4n0r <jaykumar20march@gmail.com>

35C4n0r committed:
ref: refactored files

- moved ahocode.py to a python package
- Updated metadata in setup.cfg

Signed-off-by: 35C4n0r <jaykumar20march@gmail.com>

JonoYang commented on setup.cfg:
```diff
@@ -1,12 +1,12 @@
 long_description = file:README.rst
 long_description_content_type = text/x-rst
-url = https://github.com/nexB/skeleton
+url = https://github.com/nexB/ahocde
```
```suggestion
url = https://github.com/nexB/ahocode
```

JonoYang commented on tests/test.py:
```diff
@@ -0,0 +1,78 @@
+#
```
Rename this file `test_ahocode.py`

JonoYang commented on src/ahocode/ahocode.py:
```diff
@@ -0,0 +1,371 @@
+        zero_state = self._zero_state
+        current_state = zero_state
+        offset = 0
+        if (start and end) is not None:
```
I think it is better to use `==` and `!=` operators for value comparisons instead of `is`. From this answer https://stackoverflow.com/questions/132988/is-there-a-difference-between-and-is:

> `==` is for value equality. Use it when you would like to know if two objects have the same value.
> `is` is for reference equality. Use it when you would like to know if two references refer to the same object.

Please review and revise all use of `is` in your code, unless you really are checking to see if two variables are the same object in memory.

JonoYang commented on tests/test.py:
```diff
@@ -0,0 +1,78 @@
+# ScanCode is a trademark of nexB Inc.
+# SPDX-License-Identifier: Apache-2.0
+# See http://www.apache.org/licenses/LICENSE-2.0 for the license text.
+# See https://github.com/nexB/skeleton for support or download.
```
```suggestion
# See https://github.com/nexB/ahocode for support or download.

```

JonoYang commented on tests/test.py:
```diff
@@ -0,0 +1,78 @@
+from ahocode import ahocode
+
+
+class test_automaton_methods(unittest.TestCase):
```
I see these tests are from https://github.com/WojciechMula/pyahocorasick/blob/master/tests/test_basic.py. 

Please make a comment saying that the tests are from there with a link to the code.

35C4n0r replied:
Thanks for this, i also added the proper link and repo name in ahocode.py.

JonoYang reviewed (changes requested):
@35C4n0r 

Thanks for the updates! There are a bit more changes that need to be done.

35C4n0r committed:
ref: refactored files

- replaced is comparator to ==.
- Updated metadata in setup.cfg.
- renamed test.py to test_ahocode.py.
- added credits in test_ahocode.py and ahocode.py.

Signed-off-by: 35C4n0r <jaykumar20march@gmail.com>

35C4n0r committed:
add: added credits

- added credits in test_ahocode.py and ahocode.py.

Signed-off-by: 35C4n0r <jaykumar20march@gmail.com>

35C4n0r committed:
add: added new tests

- added tests for trie methods.
- added kind property and __len__ for Automaton class

Signed-off-by: 35C4n0r <jaykumar20march@gmail.com>

JonoYang reviewed (changes requested):
@35C4n0r 

Looks great! One last thing, can you update README.rst? It currently still the same one from the skelton repo. Also, what do you think the intbitset fallback library should be named? We have `bitcode` and `Pybitset` so far.

35C4n0r commented:
> @35C4n0r 
> 
> Looks great! One last thing, can you update README.rst? It currently still the same one from the skelton repo. Also, what do you think the intbitset fallback library should be named? We have `bitcode` and `Pybitset` so far.
> 

Thanks for the review @JonoYang.I will update the readme.md accordingly. 
As for the name I like bitcode more :P

35C4n0r committed:
add: added README.rst

Signed-off-by: 35C4n0r <jaykumar20march@gmail.com>

35C4n0r commented:
@JonoYang  Added a README.rst, not sure if it's good, please review and give some suggestions.

JonoYang commented on README.rst:
```diff
@@ -1,59 +1,48 @@
-all new Python projects and is mergeable in existing repositories as well.
+ahocode
+==================
+This repo is a pure python implementation for `pyahocorasick <https://github.com/WojciechMula/pyahocorasick>`_.
```
```suggestion
This repo is a pure python implementation for `pyahocorasick <https://github.com/WojciechMula/pyahocorasick>`_, inspired by `abusix/ahocorapy <https://github.com/abusix/ahocorapy>`_.
```

JonoYang reviewed (changes requested):
@35C4n0r Readme looks good, I just have a small change. Other than that, I think we're ready to  merge.

35C4n0r committed:
Update README.rst

Co-authored-by: Jono Yang <JonoYang@users.noreply.github.com>

35C4n0r committed:
feat: updated README.rst

Signed-off-by: 35C4n0r <jaykumar20march@gmail.com>

JonoYang commented:
@35C4n0r Thanks for the update! Merging, as everything looks good.
